File: apps/restate/src/soniox.ts
import { z } from "zod";

const SONIOX_API_HOST = "https://api.soniox.com";

export class SonioxError extends Error {
  constructor(
    message: string,
    public readonly statusCode: number,
    public readonly isRetryable: boolean,
  ) {
    super(message);
    this.name = "SonioxError";
  }
}

function isRetryableStatus(status: number): boolean {
  return status === 429 || status >= 500;
}

export const SonioxCallback = z.object({
  id: z.string(),
  status: z.enum(["completed", "error"]),
});

export type SonioxCallbackType = z.infer<typeof SonioxCallback>;

type SonioxToken = {
  text: string;
  start_ms?: number;
  end_ms?: number;
  confidence?: number;
  speaker?: number | string;
};

type SonioxTranscriptResponse = {
  text: string;
  tokens: SonioxToken[];
};

export async function transcribeWithCallback(
  audioUrl: string,
  callbackUrl: string,
  apiKey: string,
): Promise<string> {
  const response = await fetch(`${SONIOX_API_HOST}/v1/transcriptions`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: "stt-async-v3",
      audio_url: audioUrl,
      webhook_url: callbackUrl,
      enable_speaker_diarization: true,
      enable_language_identification: true,
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new SonioxError(
      `Soniox: ${response.status} - ${errorText}`,
      response.status,
      isRetryableStatus(response.status),
    );
  }

  const result = (await response.json()) as { id: string };
  if (!result.id) {
    throw new Error("Soniox: missing transcription id");
  }

  return result.id;
}

export async function fetchTranscript(
  transcriptionId: string,
  apiKey: string,
): Promise<string> {
  const response = await fetch(
    `${SONIOX_API_HOST}/v1/transcriptions/${transcriptionId}/transcript`,
    {
      headers: {
        Authorization: `Bearer ${apiKey}`,
      },
    },
  );

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(
      `Soniox fetch transcript: ${response.status} - ${errorText}`,
    );
  }

  const result = (await response.json()) as SonioxTranscriptResponse;
  return result.text || renderTokens(result.tokens);
}

function renderTokens(tokens: SonioxToken[]): string {
  return tokens.map((token) => token.text).join("");
}

File: crates/restate-stt/src/soniox.rs
pub use soniox::CallbackPayload as SonioxCallback;
pub use soniox::Error as SonioxError;
pub use soniox::fetch_transcript_raw;

pub async fn transcribe_with_callback(
    client: &reqwest::Client,
    audio_url: &str,
    callback_url: &str,
    api_key: &str,
    options: Option<&serde_json::Value>,
) -> Result<String, SonioxError> {
    let mut body = serde_json::json!({
        "model": "stt-async-v4",
        "audio_url": audio_url,
        "webhook_url": callback_url,
        "enable_speaker_diarization": true,
        "enable_language_identification": true,
    });

    if let Some(overrides) = options {
        if let (Some(base), Some(extra)) = (body.as_object_mut(), overrides.as_object()) {
            for (k, v) in extra {
                base.insert(k.clone(), v.clone());
            }
        }
    }

    soniox::create_transcription(client, &body, api_key).await
}
